File: 3rd_party_slots/python_client/python_bus_client/service_bus.py
import threading
from python_bus_client.port_pool import PortPool
from python_bus_client.bus_node import BusNode
from python_bus_client.proxy import PatternMatchingQueryHandler
from python_bus_client.bus import BusCommand
from python_bus_client.logger import log


class ServiceBus:
    def __init__(self, host_id: str, known_peer: str, commands: list[str], port_lower: int, port_upper: int):
        self.service_list: list[str] = []
        self.next_request_serial: int = 0
        self.bus_node: 'BusNode' = BusNode(host_id, commands, known_peer)
        PortPool.initialize_statics(port_lower, port_upper)

    def issue_bus_command(self, proxy: 'PatternMatchingQueryHandler'):
        log.info(f"{self.bus_node.node_id} is issuing BUS command <{proxy.command}>")

        self.next_request_serial += 1
        proxy.requestor_id = self.bus_node.node_id
        proxy.serial = self.next_request_serial
        proxy.proxy_port = PortPool.get_port()

        if proxy.proxy_port == 0:
            raise RuntimeError("No port is available to start bus command proxy")
        else:
            proxy.setup_proxy_node()

            args = [
                proxy.requestor_id,
                str(proxy.serial),
                proxy.proxy_node.node_id
            ]
            args.extend(proxy.args)

            # TODO
            try:
                self.bus_node.send_bus_command(proxy.command, args)
            except Exception:
                pass


class ServiceBusSingletonMeta(type):
    """
    This is a thread-safe implementation of Singleton.
    """

    _instances = {}

    _lock = threading.Lock()

    def __call__(cls, *args, **kwargs):
        with cls._lock:
            if cls not in cls._instances:
                instance = super().__call__(*args, **kwargs)
                cls._instances[cls] = instance
        return cls._instances[cls]


class ServiceBusSingleton(metaclass=ServiceBusSingletonMeta):
    def __init__(self, host_id: str, known_peer: str, port_lower: int, port_upper: int) -> None:
        self.service_bus = ServiceBus(
            host_id,
            known_peer,
            [BusCommand.PATTERN_MATCHING_QUERY],
            port_lower,
            port_upper
        )

    def get_instance(self) -> 'ServiceBus':
        return self.service_bus

File: 3rd_party_slots/rust_metta_bus_client/src/service_bus.rs
use std::{
	sync::{Arc, Mutex, RwLock},
	thread::{self, sleep},
	time::Duration,
};

use tokio::runtime::Builder;

use crate::{
	base_proxy_query::{BaseQueryProxy, BaseQueryProxyT},
	bus_node::BusNode,
	port_pool::PortPool,
	proxy::StarNode,
	types::BoxError,
};

#[derive(Debug, Default, Clone)]
pub struct ServiceBus {
	pub service_list: Vec<String>,
	pub bus_node: Arc<Mutex<BusNode>>,
	pub next_request_serial: u64,
}

impl ServiceBus {
	pub fn new(
		host_id: String, known_peer: String, commands: Vec<String>, port_lower: u16,
		port_upper: u16,
	) -> Result<Self, BoxError> {
		PortPool::initialize_statics(port_lower, port_upper)?;
		let host_id = format!("{}:{}", host_id, PortPool::get_port());
		let bus_node =
			Arc::new(Mutex::new(BusNode::new(host_id.clone(), commands.clone(), known_peer)?));
		Ok(Self { service_list: commands, bus_node, next_request_serial: 0 })
	}

	pub fn issue_bus_command(
		&mut self, proxy_arc: Arc<Mutex<BaseQueryProxy>>,
	) -> Result<(), BoxError> {
		let mut proxy = proxy_arc.lock().unwrap();
		let locked_bus_node = self.bus_node.lock().unwrap();
		log::debug!(target: "das", "{} is issuing BUS command <{}>", locked_bus_node.node_id(), proxy.command);
		proxy.requestor_id = locked_bus_node.node_id();
		self.next_request_serial += 1;
		proxy.serial = self.next_request_serial;
		proxy.proxy_port = PortPool::get_port();
		if proxy.proxy_port == 0 {
			panic!("No port is available to start bus command proxy");
		} else {
			proxy.setup_proxy_node(proxy_arc.clone(), None, None);
			let mut args = Vec::new();
			args.push(proxy.requestor_id.clone());
			args.push(proxy.serial.to_string());
			args.push(proxy.proxy_node.node_id());
			for arg in &proxy.args {
				args.push(arg.clone());
			}
			match locked_bus_node.send_bus_command(proxy.command.clone(), args) {
				Ok(_) => (),
				Err(_) => proxy.answer_flow_finished = true,
			}
		}
		Ok(())
	}

	pub fn service(&self) -> Vec<String> {
		self.service_list.clone()
	}

	pub fn join_network_thread(&mut self) -> Result<(), BoxError> {
		let host_id_clone = self.bus_node.lock().unwrap().host_id.clone();
		let bus_clone = self.bus_node.clone();
		thread::spawn(move || {
			let runtime = Arc::new(RwLock::new(Some(Arc::new(
				Builder::new_multi_thread().enable_all().build().unwrap(),
			))));

			let proxy_arc =
				Arc::new(Mutex::new(BaseQueryProxy::default_with_runtime(runtime.clone())));

			StarNode::serve(host_id_clone.clone(), proxy_arc.clone(), runtime.clone()).unwrap();

			let mut send_join_network = true;
			loop {
				let mut bus_node = bus_clone.lock().unwrap();

				if send_join_network {
					log::debug!(target: "das", "BusNode::join_network(): Joining network with peer {}", bus_node.peer_id);
					bus_node
						.send(
							"node_joined_network".to_string(),
							vec![bus_node.host_id.clone()],
							bus_node.peer_id.clone(),
							true,
						)
						.unwrap();
					send_join_network = false;
					sleep(Duration::from_millis(250));
				}

				let proxy = proxy_arc.lock().unwrap();
				for (command, owner) in proxy.service_list.iter() {
					bus_node.bus.set_ownership(command.clone(), owner);
				}
				drop(proxy);
				drop(bus_node);

				sleep(Duration::from_millis(1_000));
			}
		});
		Ok(())
	}
}
